bookkeeper-server/src/test/java/org/apache/bookkeeper/client/MockBookieWatcher.java

```diff
@@ -0,0 +1,100 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.bookkeeper.client;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+import org.apache.bookkeeper.client.BKException.BKNotEnoughBookiesException;
+import org.apache.bookkeeper.net.BookieSocketAddress;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Mock implementation of BookieWatcher.
+ */
+public class MockBookieWatcher implements BookieWatcher {
```

Comment:
Why not just mock metadata driver (including RegistrationClient)? In that way, all BookieWatchImpl logic can actually be exercised during testing.

Reply:
LedgerHandle doesn't use registration client, doesn't get one passed in. It does get a BookieWatcher though, which is why i went for that. It's a bit messy, because placement policy is also passed in to the LedgerHandle. The relationship between these classes needs to be cleaned up a bit. For now I was just going for what was easiest.

Reply:
> LedgerHandle doesn't use registration client, doesn't get one passed in. 

sorry I don't quite get this point.

> It does get a BookieWatcher though, which is why i went for that. 

I think what you want is to provide a mock bookie watcher when you are constructing ledger handle. my comment is instead of having a mock bookie watcher, you can have a mock registration client, then you can construct a *real* watcher with mock reg client and a real policy. 

```
RegistratonClient mockRegClient = mockRegistrationClient();
EnsemblePlacementPolicy policy = new DefaultEnsemblePlacementPolicy();
BookieWatcher mockWatcher = new BookieWatcher(mockRegClient, ...);
```

I understand the motivation of providing mocks. however I don't think we should provide mock for everything, because mock means maintaining two different implementations, when you have two different implementations, they usually can not keep up with each other, then problems will start increasing when they diverge. in most of the time, we should just mock the components that interact with network (like metadata driver, bookie client). unless you need a mock implementation to control failure points, that can be an exception.

since we are doing refactor to improve testability, we should just think clearly about what we will do. having a mock metadata driver which can be used for providing a *mock* bookie watcher seems to me is a more sustainable approach.